README.md
# DevTrends.ConfigurationExtensions

[![Test Status](https://github.com/devtrends-org/configuration-extensions/actions/workflows/publish.yml/badge.svg)](https://github.com/devtrends-org/configuration-extensions/actions)

[![DevTrends.ConfigurationExtensions on NuGet](https://badgen.net/nuget/v/DevTrends.ConfigurationExtensions)](https://www.nuget.org/packages/DevTrends.ConfigurationExtensions/)

This package provides a couple of extension methods that allow you to bind configuration to records and classes with parameterized (non parameterless) constructors.

It also works nicely with nullable types and nullable reference types (NRTs) and should help you avoid some of those annoying 'possible null reference' warnings.

The package works on the 'fail fast' principle and any missing non-nullable configuration values will result in an exception at startup.

Note that the package extends the built-in configuration so it will pick up settings from all the usual sources such as appsettings, environment variables etc.

## Installation

    dotnet add package DevTrends.ConfigurationExtensions

## Usage

Given a record:

    public record Foo(string Bar, int Blah)

If you have the following in your appsettings:

    "Foo": {
      "Bar": "hello",
      "Blah": 42
    }

You can bind the record to the settings by using:

    builder.Configuration.Bind<Foo>()

You will also need to add the following namespace to program.cs:

    using DevTrends.ConfigurationExtensions;

## Dependency Injection

You will probably want to register the strongly typed configuration classes with the DI container so instead of using the above, you can use:

    builder.Services.AddSingleton(builder.Configuration.Bind<Foo>());

## What is supported?

You can bind to properties of the following type (plus nullable version including nullable reference types such as string?):

-   string
-   int
-   bool
-   decimal
-   datetime
-   uri
-   nested classes

Your classes or records must have a public constructor which sets all properties that you wish to automatically populate.

If you have any optional configuration values, be sure to use nullable types.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

## [1.2.1] - 2023-04-06

- Fixed Uri bind exception message (thanks for the PR from StMarian)

## [1.2.0] - 2022-08-08

- You can now bind to properties of type System.Uri

## [1.1.0] - 2022-05-26

- Changed library to extend IConfiguration instead of ConfigurationManager in order to support Azure Functions.

## [1.0.0] - 2021-12-06

- Initial release


ConfigurationExtensions.Tests/ConfigurationManagerExtensionsTests.cs
using System;
using System.Collections.Generic;
using Microsoft.Extensions.Configuration;
using Xunit;

namespace DevTrends.ConfigurationExtensions.Tests;

public class ConfigurationManagerExtensionsTests
{
    [Fact]
    public void NoPublicConstructor_Throws()
    {
        var configurationManager = new ConfigurationManager();

        Assert.Throws<ArgumentException>(() => configurationManager.Bind<NoPublicConstructor>());
    }

    [Fact]
    public void RequiredString_Throws_When_Config_Missing()
    {
        var configurationManager = new ConfigurationManager();

        Assert.Throws<ConfigurationBindException>(() => configurationManager.Bind<RequiredString>());
    }

    [Fact]
    public void RequiredString_Binds()
    {
        var configurationManager = new ConfigurationManager();

        configurationManager.AddInMemoryCollection(new List<KeyValuePair<string, string>>
        {
            new KeyValuePair<string, string>("Foo:Foo", "text string")
        });

        var result = configurationManager.Bind<RequiredString>("Foo");

        Assert.Equal("text string", result.Foo);
    }

    [Fact]
    public void RequiredString_Binds_Using_Inferred_Section_Name()
    {
        var configurationManager = new ConfigurationManager();

        configurationManager.AddInMemoryCollection(new List<KeyValuePair<string, string>>
        {
            new KeyValuePair<string, string>("RequiredString:Foo", "text string")
        });

        var result = configurationManager.Bind<RequiredString>();

        Assert.Equal("text string", result.Foo);
    }

    [Fact]
    public void OptionalString_Binds()
    {
        var configurationManager = new ConfigurationManager();

        configurationManager.AddInMemoryCollection(new List<KeyValuePair<string, string>>
        {
            new KeyValuePair<string, string>("OptionalString:Foo", "optional string")
        });

        var result = configurationManager.Bind<OptionalString>();

        Assert.Equal("optional string", result.Foo);
    }

    [Fact]
    public void OptionalString_Binds_Null_When_Config_Missing()
    {
        var configurationManager = new ConfigurationManager();

        var result = configurationManager.Bind<OptionalString>();

        Assert.Null(result.Foo);
    }

    [Fact]
    public void MultipleTypes_Binds()
    {
        var configurationManager = new ConfigurationManager();

        configurationManager.AddInMemoryCollection(new List<KeyValuePair<string, string>>
        {
            new KeyValuePair<string, string>("MultipleTypes:Foo", "s1"),
            new KeyValuePair<string, string>("MultipleTypes:Bar", "42"),
            new KeyValuePair<string, string>("MultipleTypes:Other", bool.TrueString),
            new KeyValuePair<string, string>("MultipleTypes:Misc", "3.14"),
            new KeyValuePair<string, string>("MultipleTypes:Blah", "1999-12-31"),
        });

        var result = configurationManager.Bind<MultipleTypes>();

        Assert.Equal("s1", result.Foo);
        Assert.Equal(42, result.Bar);
        Assert.True(result.Other);
        Assert.Equal(3.14m, result.Misc);
        Assert.Equal(DateTime.Parse("1999-12-31"), result.Blah);
    }

    [Fact]
    public void MultipleTypes_Throws_When_Missing_Int()
    {
        var configurationManager = new ConfigurationManager();

        configurationManager.AddInMemoryCollection(new List<KeyValuePair<string, string>>
        {
            new KeyValuePair<string, string>("MultipleTypes:Foo", "s1"),
            new KeyValuePair<string, string>("MultipleTypes:Other", bool.TrueString),
            new KeyValuePair<string, string>("MultipleTypes:Misc", "3.14"),
            new KeyValuePair<string, string>("MultipleTypes:Blah", "1999-12-31"),
        });

        var ex = Assert.Throws<ConfigurationBindException>(() => configurationManager.Bind<MultipleTypes>());
        Assert.Contains("Unable to set Bar", ex.Message);
    }

    [Fact]
    public void MultipleTypes_Throws_When_Missing_Bool()
    {
        var configurationManager = new ConfigurationManager();

        configurationManager.AddInMemoryCollection(new List<KeyValuePair<string, string>>
        {
            new KeyValuePair<string, string>("MultipleTypes:Foo", "s1"),
            new KeyValuePair<string, string>("MultipleTypes:Bar", "42"),
            new KeyValuePair<string, string>("MultipleTypes:Misc", "3.14"),
            new KeyValuePair<string, string>("MultipleTypes:Blah", "1999-12-31"),
        });

        var ex = Assert.Throws<ConfigurationBindException>(() => configurationManager.Bind<MultipleTypes>());
        Assert.Contains("Unable to set Other", ex.Message);
    }

    [Fact]
    public void MultipleTypes_Throws_When_Missing_Decimal()
    {
        var configurationManager = new ConfigurationManager();

        configurationManager.AddInMemoryCollection(new List<KeyValuePair<string, string>>
        {
            new KeyValuePair<string, string>("MultipleTypes:Foo", "s1"),
            new KeyValuePair<string, string>("MultipleTypes:Bar", "42"),
            new KeyValuePair<string, string>("MultipleTypes:Other", bool.TrueString),
            new KeyValuePair<string, string>("MultipleTypes:Blah", "1999-12-31"),
        });

        var ex = Assert.Throws<ConfigurationBindException>(() => configurationManager.Bind<MultipleTypes>());
        Assert.Contains("Unable to set Misc", ex.Message);
    }

    [Fact]
    public void MultipleTypes_Throws_When_Missing_DateTime()
    {
        var configurationManager = new ConfigurationManager();

        configurationManager.AddInMemoryCollection(new List<KeyValuePair<string, string>>
        {
            new KeyValuePair<string, string>("MultipleTypes:Foo", "s1"),
            new KeyValuePair<string, string>("MultipleTypes:Bar", "42"),
            new KeyValuePair<string, string>("MultipleTypes:Other", bool.TrueString),
            new KeyValuePair<string, string>("MultipleTypes:Misc", "3.14")
        });

        var ex = Assert.Throws<ConfigurationBindException>(() => configurationManager.Bind<MultipleTypes>());
        Assert.Contains("Unable to set Blah", ex.Message);
    }

    [Fact]
    public void MultipleOptionalTypes_Binds()
    {
        var configurationManager = new ConfigurationManager();

        configurationManager.AddInMemoryCollection(new List<KeyValuePair<string, string>>
        {
            new KeyValuePair<string, string>("MultipleOptionalTypes:Foo", "s1"),
            new KeyValuePair<string, string>("MultipleOptionalTypes:Bar", "42"),
            new KeyValuePair<string, string>("MultipleOptionalTypes:Other", bool.TrueString),
            new KeyValuePair<string, string>("MultipleOptionalTypes:Misc", "3.14"),
            new KeyValuePair<string, string>("MultipleOptionalTypes:Blah", "1999-12-31"),
        });

        var result = configurationManager.Bind<MultipleOptionalTypes>();

        Assert.Equal("s1", result.Foo);
        Assert.Equal(42, result.Bar);
        Assert.True(result.Other);
        Assert.Equal(3.14m, result.Misc);
        Assert.Equal(DateTime.Parse("1999-12-31"), result.Blah);
    }

    [Fact]
    public void MultipleOptionalTypes_Binds_Null_When_Config_Missing()
    {
        var configurationManager = new ConfigurationManager();

        var result = configurationManager.Bind<MultipleOptionalTypes>();

        Assert.Null(result.Foo);
        Assert.Null(result.Bar);
        Assert.Null(result.Other);
        Assert.Null(result.Misc);
        Assert.Null(result.Blah);
    }

    [Fact]
    public void Nested_NoPublicConstructor_Throws()
    {
        var configurationManager = new ConfigurationManager();

        Assert.Throws<ArgumentException>(() => configurationManager.Bind<NestedNoPublicConstructor>());
    }

    [Fact]
    public void NestedRequired_Binds()
    {
        var configurationManager = new ConfigurationManager();

        configurationManager.AddInMemoryCollection(new List<KeyValuePair<string, string>>
        {
            new KeyValuePair<string, string>("NestedRequired:Something", "hi"),
            new KeyValuePair<string, string>("NestedRequired:Other:Foo", "nested")
        });

        var result = configurationManager.Bind<NestedRequired>();

        Assert.Equal("hi", result.Something);
        Assert.NotNull(result.Other);
        Assert.Equal("nested", result.Other.Foo);
    }

    [Fact]
    public void NestedRequired_Throws_When_Config_Missing()
    {
        var configurationManager = new ConfigurationManager();

        configurationManager.AddInMemoryCollection(new List<KeyValuePair<string, string>>
        {
            new KeyValuePair<string, string>("NestedRequired:Something", "hi")
        });

        Assert.Throws<ConfigurationBindException>(() => configurationManager.Bind<NestedRequired>());
    }

    [Fact]
    public void NestedOptional_Binds()
    {
        var configurationManager = new ConfigurationManager();

        configurationManager.AddInMemoryCollection(new List<KeyValuePair<string, string>>
        {
            new KeyValuePair<string, string>("NestedOptional:Something", "hi"),
            new KeyValuePair<string, string>("NestedOptional:Other:Foo", "nested")
        });

        var result = configurationManager.Bind<NestedOptional>();

        Assert.Equal("hi", result.Something);
        Assert.NotNull(result.Other);
        Assert.Equal("nested", result.Other.Foo);
    }

    [Fact]
    public void NestedOptional_Binds_Null_When_Config_Missing()
    {
        var configurationManager = new ConfigurationManager();

        configurationManager.AddInMemoryCollection(new List<KeyValuePair<string, string>>
        {
            new KeyValuePair<string, string>("NestedOptional:Something", "hi")
        });

        var result = configurationManager.Bind<NestedOptional>();

        Assert.Equal("hi", result.Something);
        Assert.NotNull(result.Other);
        Assert.Null(result.Other.Foo);
    }

    [Fact]
    public void NestedOptionalType_Binds()
    {
        var configurationManager = new ConfigurationManager();

        configurationManager.AddInMemoryCollection(new List<KeyValuePair<string, string>>
        {
            new KeyValuePair<string, string>("NestedOptionalType:Something", "hi"),
            new KeyValuePair<string, string>("NestedOptionalType:Other:Foo", "nested")
        });

        var result = configurationManager.Bind<NestedOptionalType>();

        Assert.Equal("hi", result.Something);
        Assert.NotNull(result.Other);
        Assert.Equal("nested", result.Other!.Foo);
    }

    [Fact]
    public void NestedOptionalType_Binds_Null_When_Config_Missing()
    {
        var configurationManager = new ConfigurationManager();

        configurationManager.AddInMemoryCollection(new List<KeyValuePair<string, string>>
        {
            new KeyValuePair<string, string>("NestedOptionalType:Something", "hi")
        });

        var result = configurationManager.Bind<NestedOptionalType>();

        Assert.Equal("hi", result.Something);
        Assert.Null(result.Other);
    }
}

public class NoPublicConstructor
{
    private NoPublicConstructor()
    {
    }
}

public record RequiredString(string Foo);

public record OptionalString(string? Foo);

public record MultipleTypes(string Foo, int Bar, bool Other, decimal Misc, DateTime Blah);

public record MultipleOptionalTypes(string? Foo, int? Bar, bool? Other, decimal? Misc, DateTime? Blah);

public record NestedRequired(string Something, RequiredString Other);

public record NestedOptional(string Something, OptionalString Other);

public record NestedNoPublicConstructor(NoPublicConstructor Other);

public record NestedOptionalType(string Something, RequiredString? Other);

ConfigurationExtensions.Tests/IConfigurationExtensionsTests.cs
using System;
using System.Collections.Generic;
using Microsoft.Extensions.Configuration;
using Xunit;

namespace DevTrends.ConfigurationExtensions.Tests;

public class IConfigurationExtensionsTests
{
    [Fact]
    public void NoPublicConstructor_Throws()
    {
        var configurationManager = new ConfigurationManager();

        Assert.Throws<ArgumentException>(() => configurationManager.Bind<NoPublicConstructor>());
    }

    [Fact]
    public void RequiredString_Throws_When_Config_Missing()
    {
        var configurationManager = new ConfigurationManager();

        Assert.Throws<ConfigurationBindException>(() => configurationManager.Bind<RequiredString>());
    }

    [Fact]
    public void RequiredString_Binds()
    {
        var configurationManager = new ConfigurationManager();

        configurationManager.AddInMemoryCollection(new List<KeyValuePair<string, string>>
        {
            new KeyValuePair<string, string>("Foo:Foo", "text string")
        });

        var result = configurationManager.Bind<RequiredString>("Foo");

        Assert.Equal("text string", result.Foo);
    }

    [Fact]
    public void RequiredString_Binds_Using_Inferred_Section_Name()
    {
        var configurationManager = new ConfigurationManager();

        configurationManager.AddInMemoryCollection(new List<KeyValuePair<string, string>>
        {
            new KeyValuePair<string, string>("RequiredString:Foo", "text string")
        });

        var result = configurationManager.Bind<RequiredString>();

        Assert.Equal("text string", result.Foo);
    }

    [Fact]
    public void OptionalString_Binds()
    {
        var configurationManager = new ConfigurationManager();

        configurationManager.AddInMemoryCollection(new List<KeyValuePair<string, string>>
        {
            new KeyValuePair<string, string>("OptionalString:Foo", "optional string")
        });

        var result = configurationManager.Bind<OptionalString>();

        Assert.Equal("optional string", result.Foo);
    }

    [Fact]
    public void OptionalString_Binds_Null_When_Config_Missing()
    {
        var configurationManager = new ConfigurationManager();

        var result = configurationManager.Bind<OptionalString>();

        Assert.Null(result.Foo);
    }

    [Fact]
    public void MultipleTypes_Binds()
    {
        var configurationManager = new ConfigurationManager();

        configurationManager.AddInMemoryCollection(new List<KeyValuePair<string, string>>
        {
            new KeyValuePair<string, string>("MultipleTypes:Foo", "s1"),
            new KeyValuePair<string, string>("MultipleTypes:Bar", "42"),
            new KeyValuePair<string, string>("MultipleTypes:Other", bool.TrueString),
            new KeyValuePair<string, string>("MultipleTypes:Misc", "3.14"),
            new KeyValuePair<string, string>("MultipleTypes:Blah", "1999-12-31"),
            new KeyValuePair<string, string>("MultipleTypes:Url", "https://www.foo.com"),
        });

        var result = configurationManager.Bind<MultipleTypes>();

        Assert.Equal("s1", result.Foo);
        Assert.Equal(42, result.Bar);
        Assert.True(result.Other);
        Assert.Equal(3.14m, result.Misc);
        Assert.Equal(DateTime.Parse("1999-12-31"), result.Blah);
        Assert.Equal(new Uri("https://www.foo.com"), result.Url);
    }

    [Fact]
    public void MultipleTypes_Throws_When_Missing_Int()
    {
        var configurationManager = new ConfigurationManager();

        configurationManager.AddInMemoryCollection(new List<KeyValuePair<string, string>>
        {
            new KeyValuePair<string, string>("MultipleTypes:Foo", "s1"),
            new KeyValuePair<string, string>("MultipleTypes:Other", bool.TrueString),
            new KeyValuePair<string, string>("MultipleTypes:Misc", "3.14"),
            new KeyValuePair<string, string>("MultipleTypes:Blah", "1999-12-31"),
            new KeyValuePair<string, string>("MultipleTypes:Url", "https://www.foo.com"),
        });

        var ex = Assert.Throws<ConfigurationBindException>(() => configurationManager.Bind<MultipleTypes>());
        Assert.Contains("Unable to set Bar", ex.Message);
    }

    [Fact]
    public void MultipleTypes_Throws_When_Missing_Bool()
    {
        var configurationManager = new ConfigurationManager();

        configurationManager.AddInMemoryCollection(new List<KeyValuePair<string, string>>
        {
            new KeyValuePair<string, string>("MultipleTypes:Foo", "s1"),
            new KeyValuePair<string, string>("MultipleTypes:Bar", "42"),
            new KeyValuePair<string, string>("MultipleTypes:Misc", "3.14"),
            new KeyValuePair<string, string>("MultipleTypes:Blah", "1999-12-31"),
            new KeyValuePair<string, string>("MultipleTypes:Url", "https://www.foo.com"),
        });

        var ex = Assert.Throws<ConfigurationBindException>(() => configurationManager.Bind<MultipleTypes>());
        Assert.Contains("Unable to set Other", ex.Message);
    }

    [Fact]
    public void MultipleTypes_Throws_When_Missing_Decimal()
    {
        var configurationManager = new ConfigurationManager();

        configurationManager.AddInMemoryCollection(new List<KeyValuePair<string, string>>
        {
            new KeyValuePair<string, string>("MultipleTypes:Foo", "s1"),
            new KeyValuePair<string, string>("MultipleTypes:Bar", "42"),
            new KeyValuePair<string, string>("MultipleTypes:Other", bool.TrueString),
            new KeyValuePair<string, string>("MultipleTypes:Blah", "1999-12-31"),
            new KeyValuePair<string, string>("MultipleTypes:Url", "https://www.foo.com"),
        });

        var ex = Assert.Throws<ConfigurationBindException>(() => configurationManager.Bind<MultipleTypes>());
        Assert.Contains("Unable to set Misc", ex.Message);
    }

    [Fact]
    public void MultipleTypes_Throws_When_Missing_DateTime()
    {
        var configurationManager = new ConfigurationManager();

        configurationManager.AddInMemoryCollection(new List<KeyValuePair<string, string>>
        {
            new KeyValuePair<string, string>("MultipleTypes:Foo", "s1"),
            new KeyValuePair<string, string>("MultipleTypes:Bar", "42"),
            new KeyValuePair<string, string>("MultipleTypes:Other", bool.TrueString),
            new KeyValuePair<string, string>("MultipleTypes:Misc", "3.14"),
            new KeyValuePair<string, string>("MultipleTypes:Url", "https://www.foo.com"),
        });

        var ex = Assert.Throws<ConfigurationBindException>(() => configurationManager.Bind<MultipleTypes>());
        Assert.Contains("Unable to set Blah", ex.Message);
    }

    [Fact]
    public void MultipleTypes_Throws_When_Missing_Uri()
    {
        var configurationManager = new ConfigurationManager();

        configurationManager.AddInMemoryCollection(new List<KeyValuePair<string, string>>
        {
            new KeyValuePair<string, string>("MultipleTypes:Foo", "s1"),
            new KeyValuePair<string, string>("MultipleTypes:Bar", "42"),
            new KeyValuePair<string, string>("MultipleTypes:Other", bool.TrueString),
            new KeyValuePair<string, string>("MultipleTypes:Misc", "3.14"),
            new KeyValuePair<string, string>("MultipleTypes:Blah", "1999-12-31")
        });

        var ex = Assert.Throws<ConfigurationBindException>(() => configurationManager.Bind<MultipleTypes>());
        Assert.Contains("Unable to set Url", ex.Message);
    }

    [Fact]
    public void MultipleOptionalTypes_Binds()
    {
        var configurationManager = new ConfigurationManager();

        configurationManager.AddInMemoryCollection(new List<KeyValuePair<string, string>>
        {
            new KeyValuePair<string, string>("MultipleOptionalTypes:Foo", "s1"),
            new KeyValuePair<string, string>("MultipleOptionalTypes:Bar", "42"),
            new KeyValuePair<string, string>("MultipleOptionalTypes:Other", bool.TrueString),
            new KeyValuePair<string, string>("MultipleOptionalTypes:Misc", "3.14"),
            new KeyValuePair<string, string>("MultipleOptionalTypes:Blah", "1999-12-31"),
            new KeyValuePair<string, string>("MultipleOptionalTypes:Url", "https://www.foo.com"),
        });

        var result = configurationManager.Bind<MultipleOptionalTypes>();

        Assert.Equal("s1", result.Foo);
        Assert.Equal(42, result.Bar);
        Assert.True(result.Other);
        Assert.Equal(3.14m, result.Misc);
        Assert.Equal(DateTime.Parse("1999-12-31"), result.Blah);
        Assert.Equal(new Uri("https://www.foo.com"), result.Url);
    }

    [Fact]
    public void MultipleOptionalTypes_Binds_Null_When_Config_Missing()
    {
        var configurationManager = new ConfigurationManager();

        var result = configurationManager.Bind<MultipleOptionalTypes>();

        Assert.Null(result.Foo);
        Assert.Null(result.Bar);
        Assert.Null(result.Other);
        Assert.Null(result.Misc);
        Assert.Null(result.Blah);
        Assert.Null(result.Url);
    }

    [Fact]
    public void Nested_NoPublicConstructor_Throws()
    {
        var configurationManager = new ConfigurationManager();

        Assert.Throws<ArgumentException>(() => configurationManager.Bind<NestedNoPublicConstructor>());
    }

    [Fact]
    public void NestedRequired_Binds()
    {
        var configurationManager = new ConfigurationManager();

        configurationManager.AddInMemoryCollection(new List<KeyValuePair<string, string>>
        {
            new KeyValuePair<string, string>("NestedRequired:Something", "hi"),
            new KeyValuePair<string, string>("NestedRequired:Other:Foo", "nested")
        });

        var result = configurationManager.Bind<NestedRequired>();

        Assert.Equal("hi", result.Something);
        Assert.NotNull(result.Other);
        Assert.Equal("nested", result.Other.Foo);
    }

    [Fact]
    public void NestedRequired_Throws_When_Config_Missing()
    {
        var configurationManager = new ConfigurationManager();

        configurationManager.AddInMemoryCollection(new List<KeyValuePair<string, string>>
        {
            new KeyValuePair<string, string>("NestedRequired:Something", "hi")
        });

        Assert.Throws<ConfigurationBindException>(() => configurationManager.Bind<NestedRequired>());
    }

    [Fact]
    public void NestedOptional_Binds()
    {
        var configurationManager = new ConfigurationManager();

        configurationManager.AddInMemoryCollection(new List<KeyValuePair<string, string>>
        {
            new KeyValuePair<string, string>("NestedOptional:Something", "hi"),
            new KeyValuePair<string, string>("NestedOptional:Other:Foo", "nested")
        });

        var result = configurationManager.Bind<NestedOptional>();

        Assert.Equal("hi", result.Something);
        Assert.NotNull(result.Other);
        Assert.Equal("nested", result.Other.Foo);
    }

    [Fact]
    public void NestedOptional_Binds_Null_When_Config_Missing()
    {
        var configurationManager = new ConfigurationManager();

        configurationManager.AddInMemoryCollection(new List<KeyValuePair<string, string>>
        {
            new KeyValuePair<string, string>("NestedOptional:Something", "hi")
        });

        var result = configurationManager.Bind<NestedOptional>();

        Assert.Equal("hi", result.Something);
        Assert.NotNull(result.Other);
        Assert.Null(result.Other.Foo);
    }

    [Fact]
    public void NestedOptionalType_Binds()
    {
        var configurationManager = new ConfigurationManager();

        configurationManager.AddInMemoryCollection(new List<KeyValuePair<string, string>>
        {
            new KeyValuePair<string, string>("NestedOptionalType:Something", "hi"),
            new KeyValuePair<string, string>("NestedOptionalType:Other:Foo", "nested")
        });

        var result = configurationManager.Bind<NestedOptionalType>();

        Assert.Equal("hi", result.Something);
        Assert.NotNull(result.Other);
        Assert.Equal("nested", result.Other!.Foo);
    }

    [Fact]
    public void NestedOptionalType_Binds_Null_When_Config_Missing()
    {
        var configurationManager = new ConfigurationManager();

        configurationManager.AddInMemoryCollection(new List<KeyValuePair<string, string>>
        {
            new KeyValuePair<string, string>("NestedOptionalType:Something", "hi")
        });

        var result = configurationManager.Bind<NestedOptionalType>();

        Assert.Equal("hi", result.Something);
        Assert.Null(result.Other);
    }
}

public class NoPublicConstructor
{
    private NoPublicConstructor()
    {
    }
}

public record RequiredString(string Foo);

public record OptionalString(string? Foo);

public record MultipleTypes(string Foo, int Bar, bool Other, decimal Misc, DateTime Blah, Uri Url);

public record MultipleOptionalTypes(string? Foo, int? Bar, bool? Other, decimal? Misc, DateTime? Blah, Uri? Url);

public record NestedRequired(string Something, RequiredString Other);

public record NestedOptional(string Something, OptionalString Other);

public record NestedNoPublicConstructor(NoPublicConstructor Other);

public record NestedOptionalType(string Something, RequiredString? Other);